### PR TITLE
Add env and file options to extend support for different injection mechanisms

### PR DESCRIPTION
**Pull request by mr-beerkiss**
- ~added support for `env:` and `file:` prefixes on the `token` field to support different mechanisms of reading the token value.~
- Added support for `env` and `file` parameters to support different mechanisms of supplying the token

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,39 +3,77 @@
...
 You can also specify a custom npm registry if you are using your own mirror.
 
 ```yml
 steps:
   - command: yarn install
     plugins:
-      seek-oss/private-npm#v1.0.1:
-        token: ${NPM_TOKEN}
+      seek-oss/private-npm#v1.1.1:
+        token: ${MY_TOKEN}
```
Comment (mattcompiles) [inline README.md]: Maybe update this to the preferred approach?

File: README.md [hunk 2]
```diff
@@ -3,39 +3,77 @@
...
         registry: //myprivatenpm.com/
 ```
 
 ## Configuration
 
 ### `token` (required)
-The value of the NPM token.  
+The value of the NPM token.  Using the prefix `env:` of `file:` changes the behaviour of how the token is read.  By
```
Comment (mattcompiles) [inline README.md]: Using the prefix `env:` or `file:`

**Review by mattcompiles — COMMENTED** [review]
Looks good mate. Just a couple of typos.

File: README.md [hunk 3]
```diff
@@ -3,39 +3,77 @@
...
+To read the value from an environment variable named `MY_TOKEN` when the plugin executes, use the `env:` prefix.
 
 ```yml
 steps:
   - command: yarn install
     plugins:
-      seek-oss/private-npm#v1.0.1:
-        token: ${NPM_TOKEN}
+      seek-oss/private-npm#v1.1.1:
+        token: "env:MY_TOKEN"
```
Comment (zsims) [inline README.md]: How do you disambiguate tokens vs parameters for the token?

If you do this via YAML then you can have the best of all worlds:

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN`:
```
  seek-oss/private-npm#v1.1.1:
    token: "hard coded"
```

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN_ENV`:
```
  seek-oss/private-npm#v1.1.1:
    token:
      env: MY_TOKEN
```

`BUILDKITE_PLUGIN_PRIVATE_NPM_TOKEN_FILE`:
```
  seek-oss/private-npm#v1.1.1:
    token:
      file: "my-file"
```

Comment (mr-beerkiss) [reply README.md]: Yep that was my first instinct, but having three approaches means I can't make any one of them required in the `plugin.yml` file.  I think this is a better approach so perhaps it worth the trade off of having that extra layer of validation enforced by Buildkite

File: README.md (hunk 1 above)
Comment (mr-beerkiss) [reply README.md]: Not sure I follow you here, I think this is the variant we wish to deprecate in favour of `env` or `file`

Comment (mattcompiles) [reply README.md]: This is the section in the docs talking about custom npm registries. We should reference the non deprecated version of the API here.

Comment (mr-beerkiss) [reply README.md]: Oh right I'm with you now.  Good point I'll fix that.

File: README.md (hunk 3 above)
Comment (zsims) [reply README.md]: I still think it's worth doing. None of the plugins seem to do "deep" schemas (e.g. what's in an `array` or that `string` can't be empty). The approach in this PR relies on inspecting the opaque token.

You can also do validation at plugin execution time, e.g. that one of those three values is required

Comment (zsims) [reply README.md]: IMO it's worth doing:

```
  properties:
    registry:
      type: string
    token:
      type: [string, object]
      properties:
        env:
          type: string
        file:
          type: string
``` 

and some basic validation in bash (e.g. one required). I really don't think the approach here is consistent with other Buildkite plugins.

**Comment by mr-beerkiss** [comment]
@zsims & @mattsjones I've implemented the changes suggested by @zsims to have explicit fields for `env`, `file` and `token` behaviours.  Updated tests and docs accordingly

File: README.md [hunk 4]
```diff
@@ -3,39 +3,84 @@
...
+
+The value of the NPM token will be read from a variable which is available to the Buildkite YAML parsing context.  
+This value is interpolated when the YAML configuration is parsed by the Buildgent agent and provided to the plugin "as 
+is".
 
-Example: `${NPM_TOKEN}`
+Example: `${MY_TOKEN}`
+> **NOTE:** Don't put your tokens into source control.  Don't use web interfaces you don't control to inject them into 
+> your environment either.  Rather use a Secrets Manager.  If you are an AWS user, perhaps consider the 
+> [aws-sm-buildkite-plugin](https://github.com/seek-oss/aws-sm-buildkite-plugin) which works well with this plugin.
```
Comment (zsims) [inline README.md:79]: Nice!

**Review by zsims — APPROVED** [review]
Nice work! This is great

**Pull request merged**